build-emscripten.sh

```diff
@@ -2,52 +2,114 @@
+        cd aom-${AOM_VERSION}
+
+        emcmake cmake aom-source \
+            -DENABLE_CCACHE=1 \
+            -DAOM_TARGET_CPU=generic \
+            -DENABLE_DOCS=0 \
+            -DENABLE_TESTS=0 \
+            -DENABLE_EXAMPLES=0 \
+            -DENABLE_TESTDATA=0 \
+            -DENABLE_TOOLS=0 \
+            -DCONFIG_MULTITHREAD=0 \
+            -DCONFIG_RUNTIME_CPU_DETECT=0 \
+            -DBUILD_SHARED_LIBS=1 \
+            -DCMAKE_BUILD_TYPE=Release
+
+        emmake make -j${CORES}
+
+        cd ..
+    fi
+
+    CONFIGURE_ARGS_AOM="-DAOM_INCLUDE_DIR=${DIR}/aom-${AOM_VERSION}/aom-source -DAOM_LIBRARY=-L${DIR}/aom-${AOM_VERSION}"
+    LIBRARY_LINKER_FLAGS="$LIBRARY_LINKER_FLAGS -laom"
+    LIBRARY_INCLUDE_FLAGS="$LIBRARY_INCLUDE_FLAGS -L${DIR}/aom-${AOM_VERSION}"
 fi
 
-CONFIGURE_ARGS="-DENABLE_MULTITHREADING_SUPPORT=OFF -DWITH_GDK_PIXBUF=OFF -DWITH_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON"
-#export PKG_CONFIG_PATH="${DIR}/libde265-${LIBDE265_VERSION}"
+EXTRA_EXE_LINKER_FLAGS="-lembind"
+EXTRA_COMPILER_FLAGS=""
+if [ "$STANDALONE" = "1" ]; then
+    EXTRA_EXE_LINKER_FLAGS=""
+    EXTRA_COMPILER_FLAGS="-D__EMSCRIPTEN_STANDALONE_WASM__=1"
+fi
 
+CONFIGURE_ARGS="-DENABLE_MULTITHREADING_SUPPORT=OFF -DWITH_GDK_PIXBUF=OFF -DWITH_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DENABLE_PLUGIN_LOADING=OFF"
 emcmake cmake $CONFIGURE_ARGS \
-    -DLIBDE265_INCLUDE_DIR="${DIR}/libde265-${LIBDE265_VERSION}" \
-    -DLIBDE265_LIBRARY="-L${DIR}/libde265-${LIBDE265_VERSION}/libde265/.libs"
+    -DCMAKE_C_FLAGS="${EXTRA_COMPILER_FLAGS}" \
+    -DCMAKE_CXX_FLAGS="${EXTRA_COMPILER_FLAGS}" \
+    -DCMAKE_EXE_LINKER_FLAGS="${LIBRARY_LINKER_FLAGS} ${EXTRA_EXE_LINKER_FLAGS}" \
+    $CONFIGURE_ARGS_LIBDE265 \
+    $CONFIGURE_ARGS_AOM
+
 emmake make -j${CORES}
 
-export TOTAL_MEMORY=16777216
+LIBHEIFA="libheif/libheif.a"
+EXPORTED_FUNCTIONS=$($EMSDK/upstream/bin/llvm-nm $LIBHEIFA --format=just-symbols | grep "^heif_\|^de265_\|^aom_" | grep "[^:]$" | sed 's/^/_/' | paste -sd "," -)
 
 echo "Running Emscripten..."
-emcc libheif/libheif.so \
-     --bind \
-     --closure 0 \
-    -s NO_EXIT_RUNTIME=1 \
-    -s TOTAL_MEMORY=${TOTAL_MEMORY} \
+
+BUILD_FLAGS="-lembind -o libheif.js --pre-js pre.js --post-js post.js"
+RELEASE_BUILD_FLAGS="-O3"
+
+if [ "$STANDALONE" = "1" ]; then
+    echo "Building in standalone (non-web) build mode"
+    BUILD_FLAGS="-s STANDALONE_WASM=1 -s WASM=1 -o libheif.wasm --no-entry"
+fi
+
+if [ "$DEBUG" = "1" ]; then
+    echo "Building in debug mode"
+    RELEASE_BUILD_FLAGS="--profile -g"
+fi
+
+emcc "$LIBHEIFA" \
+    -s EXPORTED_FUNCTIONS="$EXPORTED_FUNCTIONS,_free,_malloc,_memcpy" \
     -s ALLOW_MEMORY_GROWTH=1 \
-    -s ASSERTIONS=0 \
-    -s INVOKE_RUN=0 \
-    -s DOUBLE_MODE=0 \
-    -s PRECISE_F32=0 \
-    -s DISABLE_EXCEPTION_CATCHING=1 \
-    -s USE_CLOSURE_COMPILER=0 \
-    -s LEGACY_VM_SUPPORT=1 \
-    --memory-init-file 0 \
-    -O3 \
+    -s ERROR_ON_UNDEFINED_SYMBOLS=0 \
```

Comment:
I have removed all the old flags here, I don't think they are neccesary.

Reply:
When not building to WASM, but plain JS, `--memory-init-file 0` avoids the generation of an extra file `libheif.js.mem`. Thus we may keep this to stay compatible to existing applications.

Reply:
@farindk Ah I did not even know it was possible to only emit JS, was that what the previous version had as output?

Reply:
Yes, previously, it was a single `libheif.js` file. Thus, we have to still support this that other software that depends on this, continues to work.

Reply:
@farindk Shouldn't `USE_WASM` have a default value of `0` then?

Reply:
Good point. But I think in this case, I would take the risk to change the default because WASM is clearly the future and it doesn't change anything for the end-user. Just of the ones building new libheif.js files.
I am currently thinking about even doing larger changes, so it will be worth for them having a look anyways.

Reply:
I have notified the following repo of the change: https://github.com/catdad-experiments/libheif-emscripten/issues/19

What kind of other changes do you want to make? I think it might be interesting to look at thread support (for the browser it's possible).